src/js/streams.ts

```diff
@@ -0,0 +1,385 @@
+type InFuncType = () =>
+  | null
+  | undefined
+  | string
+  | ArrayBuffer
+  | ArrayBufferView;
+
+// To define the output behavior of a tty we need to define put_char and fsync.
+// fsync flushes the stream.
+//
+// If isatty is false, we ignore fsync and use put_char.bind to fill in a dummy
+// value for the tty argument. We don't ever use the tty argument.
+type PutCharType = {
+  put_char: (tty: void, val: number) => void;
+  fsync: (tty: void) => void;
+};
+
+// A tty needs both a GetChar function and a PutChar pair.
+type TtyOps = {
+  get_char: GetCharType;
+} & PutCharType;
+
+/**
+ * We call refreshStreams at the end of every update method, but refreshStreams
+ * won't work until initializeStreams is called. So when INITIALIZED is false,
+ * refreshStreams is a no-op.
+ * @private
+ */
+let INITIALIZED = false;
+
+// These can't be used until they are initialized by initializeStreams.
+const ttyout_ops = {} as TtyOps;
+const ttyerr_ops = {} as TtyOps;
+const isattys = {} as {
+  stdin: boolean;
+  stdout: boolean;
+  stderr: boolean;
+};
+
+function refreshStreams() {
+  if (!INITIALIZED) {
+    return;
+  }
+  FS.unlink("/dev/stdin");
+  FS.unlink("/dev/stdout");
+  FS.unlink("/dev/stderr");
+  if (isattys.stdin) {
+    FS.symlink("/dev/tty", "/dev/stdin");
+  } else {
+    FS.createDevice("/dev", "stdin", ttyout_ops.get_char);
+  }
+  if (isattys.stdout) {
+    FS.symlink("/dev/tty", "/dev/stdout");
+  } else {
+    FS.createDevice(
+      "/dev",
+      "stdout",
+      null,
+      ttyout_ops.put_char.bind(undefined, undefined),
+    );
+  }
+  if (isattys.stderr) {
+    FS.symlink("/dev/tty", "/dev/stderr");
+  } else {
+    FS.createDevice(
+      "/dev",
+      "stderr",
+      null,
+      ttyerr_ops.put_char.bind(undefined, undefined),
+    );
+  }
+
+  // Refresh std streams so they use our new versions
+  FS.closeStream(0 /* stdin */);
+  FS.closeStream(1 /* stdout */);
+  FS.closeStream(2 /* stderr */);
+  FS.open("/dev/stdin", 0 /* write only */);
+  FS.open("/dev/stdout", 1 /* read only */);
+  FS.open("/dev/stderr", 1 /* read only */);
+}
+
+/**
+ * This is called at the end of loadPyodide to set up the streams. If
+ * loadPyodide has been given stdin, stdout, stderr arguments they are provided
+ * here. Otherwise, we set the default behaviors. This also fills in the global
+ * state in this file.
+ * @param stdin
+ * @param stdout
+ * @param stderr
+ * @private
+ */
+API.initializeStreams = function (
+  stdin?: InFuncType,
+  stdout?: (a: string) => void,
+  stderr?: (a: string) => void,
+) {
+  let stdin_isatty = false;
+  if (stdin) {
+    setStdin(stdin, stdin_isatty);
+  } else {
+    setDefaultStdin();
+  }
+  if (stdout) {
+    setStdout(stdout);
+  } else {
+    setDefaultStdout();
+  }
+
+  if (stderr) {
+    setStderr(stderr);
+  } else {
+    setDefaultStderr();
+  }
+  const ttyout_dev = FS.makedev(5, 0);
+  const ttyerr_dev = FS.makedev(6, 0);
```

Comment:
Is there a special meaning in a number 5, 6? (I don't know much about device id conventions)

Reply:
These are the same values that Emscripten uses:
https://github.com/emscripten-core/emscripten/blob/main/src/library_fs.js#L1346-L1347
I am not sure why. If we use different ones, everything still works fine.

Reply:
These numbers look like they are pretty stable in Emscripten -- they haven't been changed for at least 10 years.